test/Microsoft.Azure.WebJobs.Host.EndToEndTests/ServiceBusSessionsEndToEndTests.cs

```diff
@@ -0,0 +1,571 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.ServiceBus;
+using Microsoft.Azure.ServiceBus.Core;
+using Microsoft.Azure.WebJobs.Host.TestCommon;
+using Microsoft.Azure.WebJobs.ServiceBus;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace Microsoft.Azure.WebJobs.Host.EndToEndTests
+{
+    public class ServiceSessionsBusEndToEndTests : IDisposable
+    {
+        private const string _prefix = "core-test-";
+        private const string _queueName = _prefix + "queue1-sessions";
+        private const string _topicName = _prefix + "topic1-sessions";
+        private const string _subscriptionName = "sub1-sessions";
+        private static EventWaitHandle _waitHandle1;
+        private static EventWaitHandle _waitHandle2;
+        private readonly RandomNameResolver _nameResolver;
+        private const int SBTimeout = 120 * 1000;
+        private readonly string _connectionString;
+
+        public ServiceSessionsBusEndToEndTests()
+        {
+            var config = new ConfigurationBuilder()
+                .AddEnvironmentVariables()
+                .AddTestSettings()
+                .Build();
+
+            _connectionString = config.GetConnectionString(ServiceBus.Constants.DefaultConnectionStringName);
+
+            _nameResolver = new RandomNameResolver();
+
+            Cleanup().GetAwaiter().GetResult();
+        }
+
+        [Fact]
+        public async Task ServiceBusSessionQueue_OrderGuaranteed()
+        {
+            using (var host = ServiceBusSessionsTestHelper.CreateHost<ServiceBusSessionsTestJobs1>(_nameResolver))
+            {
+                await host.StartAsync();
+
+                _waitHandle1 = new ManualResetEvent(initialState: false);
+
+                await ServiceBusSessionsTestHelper.WriteQueueMessage(_connectionString, _queueName, "message1", "test-session");
```

Comment:
You're writing the messages serially here. Wouldn't they be processed in order even without sessions in this case? Might be better to do all these adds in parallel (e.g. create the tasks and Task.WhenAll them)?

Reply:
Without sessions order is not guaranteed. So we need to write the messages serially here.